packages/block-editor/src/components/use-block-drop-zone/index.js

```diff
@@ -125,40 +170,58 @@ export default function useBlockDropZone( {
 		[ targetRootClientId ]
 	);
 
-	const { getBlockListSettings } = useSelect( blockEditorStore );
+	const { getBlockListSettings, getBlocks } = useSelect( blockEditorStore );
 	const { showInsertionPoint, hideInsertionPoint } =
 		useDispatch( blockEditorStore );
 
-	const onBlockDrop = useOnBlockDrop( targetRootClientId, targetBlockIndex );
+	const onBlockDrop = useOnBlockDrop( targetRootClientId, dropTarget.index, {
+		operation: dropTarget.operation,
+	} );
 	const throttled = useThrottle(
-		useCallback( ( event, currentTarget ) => {
-			const blockElements = Array.from( currentTarget.children ).filter(
-				// Ensure the element is a block. It should have the `wp-block` class.
-				( element ) => element.classList.contains( 'wp-block' )
-			);
-			const targetIndex = getNearestBlockIndex(
-				blockElements,
-				{ x: event.clientX, y: event.clientY },
-				getBlockListSettings( targetRootClientId )?.orientation
-			);
-
-			setTargetBlockIndex( targetIndex === undefined ? 0 : targetIndex );
-
-			if ( targetIndex !== undefined ) {
-				const nextBlock = blockElements[ targetIndex ];
-				const previousBlock = blockElements[ targetIndex - 1 ];
-
-				// Don't show the insertion point when it's near an empty paragraph block.
-				if (
-					isEmptyParagraph( nextBlock ) ||
-					isEmptyParagraph( previousBlock )
-				) {
+		useCallback(
+			( event, currentTarget ) => {
+				const blockElements = Array.from(
+					currentTarget.children
+				).filter(
+					// Ensure the element is a block. It should have the `wp-block` class.
+					( element ) => element.classList.contains( 'wp-block' )
+				);
+				// The second value in the tuple is only needed afterwards but we don't want to recalculate it.
+				// eslint-disable-next-line @wordpress/no-unused-vars-before-return
+				const [ nearestBlockIndex, insertPosition ] =
+					getNearestBlockIndex(
+						blockElements,
+						{ x: event.clientX, y: event.clientY },
+						getBlockListSettings( targetRootClientId )?.orientation
+					);
+
+				// The block list is empty, don't show the insertion point but still allow dropping.
+				if ( nearestBlockIndex === undefined ) {
+					setDropTarget( {
+						index: 0,
+						operation: 'insert',
+					} );
 					return;
 				}
 
-				showInsertionPoint( targetRootClientId, targetIndex );
-			}
-		}, [] ),
+				const blocks = getBlocks( targetRootClientId );
+				const [ targetIndex, operation ] =
+					getDropTargetIndexAndOperation(
+						nearestBlockIndex,
+						insertPosition,
+						blocks
+					);
+
+				setDropTarget( {
+					index: targetIndex,
+					operation,
+				} );
+				showInsertionPoint( targetRootClientId, targetIndex, {
+					operation,
+				} );
```

Comment:
We would still need to extract the content to pass to `showInsertionPoint` though. And I personally think that a tuple style is more explicit and readable than returning an arbitrary object. It will not be an issue if it's fully typed though so I don't really have a strong preference 😅 .